src/lib.rs

```diff
@@ -240,3 +242,59 @@ pub struct SigElement {
     pub start: usize,
     pub end: usize,
 }
+
+#[derive(Debug, Clone, RustcDecodable, RustcEncodable)]
+pub struct Borrows {
+    pub ref_id: Id,
+    pub assignments: Vec<Assignment>,
+    pub loans: Vec<Loan>,
+    pub moves: Vec<Move>,
+}
+
+#[derive(Debug, RustcDecodable, RustcEncodable, Clone, Copy)]
+pub enum BorrowKind {
+    ImmBorrow,
+    MutBorrow,
+}
+
+#[derive(Debug, Clone, RustcDecodable, RustcEncodable)]
+pub enum CauseData {
+    ClosureCapture(SpanData),
+    AddrOf,
+    AutoRef,
+    AutoUnsafe,
+    RefBinding,
+    OverloadedOperator,
+    ClosureInvocation,
+    ForLoop,
+    MatchDiscriminant
+}
+
+#[derive(Debug, Clone, RustcDecodable, RustcEncodable)]
+pub struct Loan {
+    pub ref_id: Id,
```

Comment:
Sorry, I am responding while too tired and was also thinking about the front-end implementation too much. I had intended to mean the variable declaration since the prototypes only tracked borrows and moves from a variable declaration selected by the cursor. I hadn't attempted to show borrows of borrows yet (mostly because I hadn't figured out how to cleanly visualize that). Here the ref_ids point to the value that is being borrowed. I'm about to head to bed so I'll add examples tomorrow morning (US Eastern morning).